### PR TITLE
[ATen][CUDA][AMP] Fix dtype mismatch in linalg_vector_norm

### LINKED ISSUE #125174: [CUDA][AMP] Size-1 (scalar) norms are broken on CUDA + AMP following #122143
### 🐛 Describe the bug

A corner case of AMP + norm seems broken after #122143
```
import torch

inp = torch.randn(1, device='cuda', dtype=torch.half)
with torch.autocast('cuda', dtype=torch.float16):
    torch.linalg.vector_norm(inp)
```
CC @Aidyn-A who is working on a fix
@min-jean-cho 
@nWEIdia @tinglvv 

### Versions

Current nightlies

cc @ptrblck @mcarilli @leslie-fang-intel @jgong5

### PR DESCRIPTION
**Pull request by Aidyn-A**
Fixes #125174
cc @jianyuh @nikitaved @pearu @mruberry @walterddr @xwang233 @Lezcano @mcarilli @ptrblck @leslie-fang-intel @jgong5 @eqy @nWEIdia @tinglvv

### CODE REVIEW

**Review by eqy — APPROVED** [review]

File: aten/src/ATen/native/LinearAlgebra.cpp [hunk 1]
```diff
@@ -2839,11 +2839,20 @@ TORCH_IMPL_FUNC(linalg_vector_norm_out)(const Tensor& self, const Scalar& scalar
...
+      self_ = self;
+    }
     if (ord != 0.0) {
-      keepdim ? at::abs_outf(self, const_cast<Tensor&>(result)) : at::abs_outf(self.squeeze(reduce_dim), const_cast<Tensor&>(result));
+      keepdim ? at::abs_outf(self_, const_cast<Tensor&>(result)) : at::abs_outf(self_.squeeze(reduce_dim), const_cast<Tensor&>(result));
     } else {
-      keepdim ? at::ne_outf(self, 0, const_cast<Tensor&>(result)) : at::ne_outf(self.squeeze(reduce_dim), 0, const_cast<Tensor&>(result));
+      keepdim ? at::ne_outf(self_, 0, const_cast<Tensor&>(result)) : at::ne_outf(self_.squeeze(reduce_dim), 0, const_cast<Tensor&>(result));
     }
+    self.copy_(self_);
```
Comment (eqy) [inline aten/src/ATen/native/LinearAlgebra.cpp:2855]: Is this allowed if we have `const& self`

File: aten/src/ATen/native/LinearAlgebra.cpp [hunk 2]
```diff
@@ -2839,11 +2839,18 @@ TORCH_IMPL_FUNC(linalg_vector_norm_out)(const Tensor& self, const Scalar& scalar
...
```
Comment (lezcano) [inline aten/src/ATen/native/LinearAlgebra.cpp:2853]: ??
The returned tensor is `result`. Here you are potentially changing the input tensor!

**Review by lezcano — CHANGES_REQUESTED** [review]

**Comment by Aidyn-A** [comment]
I believe the test failure is unrelated. @lezcano, can you take another look?

**Review by lezcano — APPROVED** [review]
This is more reasonable. Thank you!

**Pull request closed without merging**